File: LeetCode/1707. Maximum XOR With an Element From Array/solution.py
class Node:
    def __init__(self):
        self.child = [None, None]

class Solution:
    def maximizeXor(self, nums: List[int], queries: List[List[int]]) -> List[int]:
        nums.sort()
        idx = sorted([i for i in range(len(queries))], key=lambda x: queries[x][-1])
        root = Node()
        pos = 0
        ans = [-1]*len(queries)
        def insert(val):
            cur = root
            for bit in range(32)[::-1]:
                flag = 0
                if val&(1<<bit) > 0:
                    flag = 1
                if not cur.child[flag]:
                    cur.child[flag] = Node()
                cur = cur.child[flag]
            return
        def query(val):
            res = 0
            cur = root
            for bit in range(32)[::-1]:
                flag = 0
                if val&(1<<bit) > 0:
                    flag = 1
                if cur.child[1-flag]:
                    res = (res<<1) + 1
                    cur = cur.child[1-flag]
                elif cur.child[flag]:
                    res = (res<<1) + 0
                    cur = cur.child[flag]
                else:
                    return -1
            return res

        for i in idx:
            x, m = queries[i]
            while pos < len(nums) and nums[pos] <= m:
                insert(nums[pos])
                pos += 1
            tmp = query(x)
            ans[i] = tmp
        return ans 

File: LeetCode/1707. Maximum XOR With an Element From Array/solution.cpp
class Solution {
public:
    vector<int> maximizeXor(vector<int>& nums, vector<vector<int>>& queries) {
        int n = nums.size(), m = queries.size();
        vector<int> ans(m,-1);
        vector<pair<int,int>> pp;
        vector<vector<int>> trie(1,vector<int>(2,-1));
        int tail = 0;
        auto insert = [&](int x){
            int cur = 0;
            for (int i = 29; i >= 0; i--){
                int idx = (x>>i)&1;
                if (trie[cur][idx] == -1){
                    trie[cur][idx] = trie.size();
                    trie.push_back({-1,-1});
                }
                cur = trie[cur][idx];
            }
        };

        auto query = [&](int y){
            if (trie.size() == 1)
                return -1;
            int cur = 0, q = 0;
            for (int i = 29; i >= 0; i--){
                int idx = ((y>>i)&1)^1;
                if (trie[cur][idx] == -1)
                    idx = 1^idx;
                cur = trie[cur][idx];
                q = (q<<1)|(idx^((y>>i)&1));
            }
            return q;
        };

        sort(nums.begin(), nums.end());
        for (int i = 0; i < m; i++)
            pp.push_back({queries[i][1],i});
        sort(pp.begin(), pp.end());
        int i = 0;
        for (auto [mx, j] : pp){
            bool flag = true;
            for (;i < n && nums[i] <= mx; i++){
                insert(nums[i]);
                flag = false;
            }
            if (flag)
                break;
            ans[j] = query(queries[j][0]);
        }
        return ans;
    }
};